From 58cb7c412c8dd09dc32a6c2e0d41f1d776ae292b Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Mon, 18 Aug 2025 17:34:44 +0800
Subject: [PATCH 115/123] mtk: mt76: mt7996: avoid setting channel multiple
 times for radar required case

In the current flow, the channel is set multiple times based on the
number of vif when radar is required. Therefore, a flag is added to
prevent this.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt76.h        | 1 +
 mt7996/main.c | 7 +++----
 mt7996/mcu.c  | 2 ++
 3 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/mt76.h b/mt76.h
index 7ec4af3b..5d9ddece 100644
--- a/mt76.h
+++ b/mt76.h
@@ -1137,6 +1137,7 @@ struct mt76_phy {
 	int tokens;
 
 	bool rdd_txq_paused;
+	bool radar_bypass_cal;
 };
 
 struct mt76_dev {
diff --git a/mt7996/main.c b/mt7996/main.c
index 172ec9dd..6eeb4798 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1169,14 +1169,13 @@ mt7996_post_channel_switch(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 	struct cfg80211_chan_def *chandef = &link_conf->chanreq.oper;
 	struct mt7996_dev *dev = mt7996_hw_dev(hw);
 	struct mt7996_phy *phy = mt7996_band_phy(dev, chandef->chan->band);
-	int ret;
 
 	mt7996_mcu_rdd_resume_tx(phy);
 
-	ret = cfg80211_chandef_dfs_required(hw->wiphy, chandef, NL80211_IFTYPE_AP);
-	if (ret <= 0)
-		return ret;
+	if (!phy->mt76->radar_bypass_cal)
+		return 0;
 
+	phy->mt76->radar_bypass_cal = false;
 	return mt76_set_channel(phy->mt76, chandef, false);
 }
 
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index bf26900c..08bc6d4d 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -5737,6 +5737,8 @@ int mt7996_mcu_set_chan_info(struct mt7996_phy *phy, u16 tag, bool sta)
 		req.center_ch2 = ieee80211_frequency_to_channel(freq2);
 	}
 
+	phy->mt76->radar_bypass_cal = req.switch_reason == CH_SWITCH_DFS;
+
 	return mt76_mcu_send_msg(&dev->mt76, MCU_WMWA_UNI_CMD(CHANNEL_SWITCH),
 				 &req, sizeof(req), true);
 }
-- 
2.45.2

